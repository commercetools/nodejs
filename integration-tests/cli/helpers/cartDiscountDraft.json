{
  "name": {
    "en": "integration-test-cart-discount"
  },
  "value": {
    "type": "relative",
    "permyriad": 1000
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "target": {
    "type": "lineItems",
    "predicate": "1 = 1"
  },
  "sortOrder": "0.1",
  "requiresDiscountCode": true
}
