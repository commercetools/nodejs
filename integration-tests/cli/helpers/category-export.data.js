export default [
  {
    key: 'fooKey',
    name: {
      en: 'categoryA',
    },
    slug: {
      en: 'category-a',
    },
    description: {
      en: 'descA',
    },
    externalId: 'a',
    assets: [],
  },
  {
    key: 'barKey',
    name: {
      en: 'categoryB',
    },
    slug: {
      en: 'category-b',
    },
    description: {
      en: 'descB',
    },
    externalId: 'b',
    assets: [],
  },
  {
    key: 'fooBarKey',
    name: {
      en: 'categoryC',
    },
    slug: {
      en: 'category-c',
    },
    description: {
      en: 'descC',
    },
    externalId: 'c',
    assets: [],
  },
]
