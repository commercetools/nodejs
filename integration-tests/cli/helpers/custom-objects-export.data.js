export default [
  {
    container: 'Ludus',
    key: 'copperKey',
    value: {
      paymentMethod: 'Cash',
      paymentID: '1',
    },
  },
  {
    container: 'Frobozz',
    key: 'jadeKey',
    value: {
      paymentMethod: 'cc',
      paymentID: '2',
    },
  },
  {
    container: 'Syrinx',
    key: 'crystalKey',
    value: {
      paymentMethod: 'new',
      paymentID: '3',
    },
  },
]
