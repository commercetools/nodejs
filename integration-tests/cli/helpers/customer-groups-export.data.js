export default [
  {
    groupName: 'Ludus',
    key: 'copperKey',
  },
  {
    groupName: 'Frobozz',
  },
  {
    groupName: 'Syrinx',
    key: 'crystalKey',
  },
]
