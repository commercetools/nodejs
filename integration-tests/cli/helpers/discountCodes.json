[{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "IT8MKKZX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlN85n0"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDHzsVX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlb6Z1u"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITfWgUVd"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITpQgE1P"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITL2pvbZ"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlrnkj5"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDJO3LT"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITuIcDI2"
}]
