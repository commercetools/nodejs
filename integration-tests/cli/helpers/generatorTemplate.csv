name.en,name.de,description.en,description.de,cartPredicate,isActive,maxApplications,maxApplicationsPerCustomer
Sammy,Valerian,greatest promo,super angebot,some cart predicate,true,9,4
