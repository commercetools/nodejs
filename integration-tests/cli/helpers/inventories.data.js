export const inventories = [{
  sku: '12345',
  quantityOnStock: 20,
  custom: {
    type: {
      key: 'custom-type',
    },
    fields: {
      description: 'integration tests!! arrgggh',
    },
  },
}]

export const customFields = [{
  key: 'custom-type',
  name: {
    en: 'customized fields',
  },
  description: {
    en: 'custom description',
  },
  resourceTypeIds: ['inventory-entry'],
  fieldDefinitions: [{
    name: 'description',
    type: {
      name: 'String',
    },
    required: false,
    label: {
      en: 'owner',
    },
    inputHint: 'SingleLine',
  }],
}]
