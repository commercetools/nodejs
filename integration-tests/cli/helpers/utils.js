import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import fetch from 'node-fetch'

export function clearData(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }
  return client.process(request, payload => {
    // Built-in states cannot be deleted
    const results =
      entityName === 'states'
        ? payload.body.results.filter(state => state.builtIn === false)
        : payload.body.results
    return Promise.all(
      results.map(result =>
        client.execute({
          uri: service
            .byId(result.id)
            .withVersion(result.version)
            .build(),
          method: 'DELETE',
        })
      )
    )
  })
}

export function createData(apiConfig, entityName, data, id) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })
  const requestOption = { projectKey: apiConfig.projectKey }
  const service = createRequestBuilder(requestOption)[entityName]
  return Promise.all(
    data.map(_data => {
      if (id) service.byId(id)
      const request = {
        uri: service.build(),
        method: 'POST',
        body: _data,
      }
      return client.execute(request)
    })
  )
}

export function getId(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }

  return client
    .execute(request)
    .then(result => Promise.resolve(result.body.results[0].id))
}
