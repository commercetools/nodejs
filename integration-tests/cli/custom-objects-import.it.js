import CustomObjectsImporter from '@commercetools/custom-objects-importer'
import { getCredentials } from '@commercetools/get-credentials'
import cloneDeep from 'lodash.clonedeep'

import allCustomObjects from './helpers/custom-objects-export.data'
import { clearData } from './helpers/utils'

let projectKey
if (process.env.CI === 'true') projectKey = 'custom-objects-import-int-test'
else projectKey = process.env.npm_config_projectkey

describe('Custom Object tests', () => {
  let apiConfig
  let customObjects

  const logger = {
    error: () => {},
    warn: () => {},
    info: () => {},
    debug: () => {},
  }

  beforeAll(async () => {
    const credentials = await getCredentials(projectKey)
    apiConfig = {
      host: 'https://auth.sphere.io',
      apiUrl: 'https://api.sphere.io',
      projectKey,
      credentials,
    }
  }, 15000)

  describe('Custom Objects Importer', () => {
    let objectImport

    beforeEach(() => {
      objectImport = new CustomObjectsImporter({ apiConfig }, logger)
      customObjects = cloneDeep(allCustomObjects)
    })

    afterEach(async () => {
      await clearData(apiConfig, 'customObjects')
    }, 15000)

    it(
      'should import custom objects to CTP',
      async () => {
        await objectImport.run(customObjects)
        const summary = objectImport.summaryReport()
        expect(summary).toMatchSnapshot()
      },
      15000
    )

    it(
      'should update custom objects on the CTP',
      async () => {
        const oldCustomObjectsToUpdate = customObjects.map(object => ({
          ...object,
          value: { paymentMethod: 'gold' },
        }))
        await objectImport.run(oldCustomObjectsToUpdate)

        // Reset the summary
        objectImport._initiateSummary()

        await objectImport.run(customObjects)
        const summary = objectImport.summaryReport()
        expect(summary).toMatchSnapshot()
      },
      15000
    )

    it(
      'should stop import on first errors by default',
      async () => {
        // Set batchSize to 1 so it executes serially
        objectImport = new CustomObjectsImporter(
          { apiConfig, batchSize: 1 },
          logger
        )

        // Make last two objects invalid
        customObjects[1].key = ''
        customObjects[2].container = ''

        try {
          await objectImport.run(customObjects)
        } catch (e) {
          // should create first object
          expect(e.summary.createdCount).toBe(1)
          // should stop after first error
          expect(e.summary.errors).toHaveLength(1)
          expect(e.summary).toMatchSnapshot()
        }
      },
      15000
    )

    it(
      'should continueOnProblems if `continueOnProblems`',
      async () => {
        // Set batchSize to 1 so it executes serially
        objectImport = new CustomObjectsImporter(
          {
            apiConfig,
            batchSize: 1,
            continueOnProblems: true,
          },
          logger
        )

        // Make two objects invalid
        customObjects[0].key = ''
        customObjects[1].container = ''

        await objectImport.run(customObjects)
        const summary = objectImport.summaryReport()
        expect(summary).toMatchSnapshot()
        const errors = summary.detailedSummary.errors
        expect(errors).toMatchSnapshot()
      },
      15000
    )
  })
})
