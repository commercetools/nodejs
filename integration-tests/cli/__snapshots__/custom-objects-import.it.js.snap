// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom Object tests Custom Objects Importer should continueOnProblems if \`continueOnProblems\` 1`] = `
Object {
  "detailedSummary": Object {
    "createErrorCount": 2,
    "createdCount": 1,
    "errors": Array [
      "key must at least have one character",
      "container must at least have one character",
    ],
    "unchangedCount": 0,
    "updateErrorCount": 0,
    "updatedCount": 0,
  },
  "reportMessage": "Summary: there were 1 successfully imported custom objects. 1 were newly created, 0 were updated and 0 were unchanged. 2 errors occurred (2 create errors and 0 update errors.)",
}
`;

exports[`Custom Object tests Custom Objects Importer should continueOnProblems if \`continueOnProblems\` 2`] = `
Array [
  "key must at least have one character",
  "container must at least have one character",
]
`;

exports[`Custom Object tests Custom Objects Importer should import custom objects to CTP 1`] = `
Object {
  "detailedSummary": Object {
    "createErrorCount": 0,
    "createdCount": 3,
    "errors": Array [],
    "unchangedCount": 0,
    "updateErrorCount": 0,
    "updatedCount": 0,
  },
  "reportMessage": "Summary: there were 3 successfully imported custom objects. 3 were newly created, 0 were updated and 0 were unchanged.",
}
`;

exports[`Custom Object tests Custom Objects Importer should stop import on first errors by default 1`] = `
Object {
  "createErrorCount": 1,
  "createdCount": 1,
  "errors": Array [
    "key must at least have one character",
  ],
  "unchangedCount": 0,
  "updateErrorCount": 0,
  "updatedCount": 0,
}
`;

exports[`Custom Object tests Custom Objects Importer should update custom objects on the CTP 1`] = `
Object {
  "detailedSummary": Object {
    "createErrorCount": 0,
    "createdCount": 0,
    "errors": Array [],
    "unchangedCount": 0,
    "updateErrorCount": 0,
    "updatedCount": 3,
  },
  "reportMessage": "Summary: there were 3 successfully imported custom objects. 0 were newly created, 3 were updated and 0 were unchanged.",
}
`;
